Refactor OL operator

edavidj opened the pull request:
**What this PR does / why we need it?**:

- Point to generic operator common libraries
- Add images, applications, cert-manager, and service binding changes to OL operator
- Update role files

**Does this PR introduce a user-facing change?**
<!--
If this PR introduces a user-facing change, it must include sufficient documentation to explain the use of the new or updated feature in addition to a summary of the change and link to the pull request.
-->
- [ ] User guide
- [ ] `CHANGELOG.md`

**Which issue(s) this PR fixes**:
<!--
Automatically closes the linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.

If you don't want any issue to get closed when this PR is merged,
then add `Fixes #<issue number>` in a comment instead and remove the next line.
-->
Fixes #119

edavidj committed:
Refactor OL operator and generate related files

edavidj committed:
refresh vendor folder entirely

edavidj committed:
Update controller with imagev1 changes

edavidj committed:
Fix positioning of imagestream changes

edavidj committed:
Add Cert manager changes to controller and test

edavidj committed:
Add additional recent PRs from runtime

edavidj committed:
update travis

edavidj committed:
Update RBAC files to include changes

edavidj committed:
Update scheme additions in controller.go

edavidj committed:
Include applications in role.yaml

edavidj committed:
Remove vendor folder cache from travis file

edavidj committed:
Update role files and include applications in controller.go

edavidj committed:
Fix issues with generate target in makefile

edavidj committed:
Re-enable e2e tests for common features

leochr commented:
@navidsh @arturdzm Please review when you get a moment. Thanks.

arturdzm reviewed (commented):
```
	predSubResource := predicate.Funcs{
		UpdateFunc: func(e event.UpdateEvent) bool {
			// Ignore updates to CR status in which case metadata.Generation does not change
			return e.MetaOld.GetGeneration() != e.MetaNew.GetGeneration() && (isClusterWide || watchNamespacesMap[e.MetaOld.GetNamespace()])
		}
```
This needs to be updated, generation check might prevent certificate status notifications to be delievered

arturdzm commented on deploy/role.yaml:
```diff
@@ -84,3 +73,21 @@ rules:
   - services
   verbs:
   - '*'
+- apiGroups:
```
- routes/custom-host is missing in this file

navidsh commented on pkg/apis/openliberty/v1beta1/openlibertyapplication_types.go:
```diff
@@ -42,10 +43,15 @@ type OpenLibertyApplicationSpec struct {
 	// +listMapKey=name
-	InitContainers []corev1.Container                    `json:"initContainers,omitempty"`
-	Serviceability *OpenLibertyApplicationServiceability `json:"serviceability,omitempty"`
+	InitContainers []corev1.Container `json:"initContainers,omitempty"`
```
(not directly related to this line)
@arturdzm do you know why the following is different from Runtime Component operator:
```
	// +listType=map		// +listType=map
	// +listMapKey=name		// +listMapKey=name
	VolumeMounts        []corev1.VolumeMount          `json:"volumeMounts,omitempty"`
```
vs
```
	// +listType=atomic
	VolumeMounts        []corev1.VolumeMount         `json:"volumeMounts,omitempty"`
```

edavidj committed:
Update role.yaml and fix predSubResource update

navidsh commented on pkg/apis/openliberty/v1beta1/openlibertyapplication_types.go:
```diff
@@ -65,12 +71,16 @@ type OpenLibertyApplicationService struct {
-	Provides *ServiceBindingProvides  `json:"provides,omitempty"`
+	Consumes             []ServiceBindingConsumes `json:"consumes,omitempty"`
+	Provides             *ServiceBindingProvides  `json:"provides,omitempty"`
+	Certificate          *Certificate             `json:"certificate,omitempty"`
```
Missing `// +k8s:openapi-gen=true`

navidsh commented on pkg/apis/openliberty/v1beta1/openlibertyapplication_types.go:
```diff
@@ -597,6 +688,47 @@ func (s *OpenLibertyApplicationStatus) SetCondition(c common.StatusCondition) {
 }
 
+// GetAnnotations returns route annotations
+func (r *OpenLibertyApplicationRoute) GetAnnotations() map[string]string {
```
Please ensure `SetCondition` code matches RC operator

navidsh commented on pkg/apis/openliberty/v1beta1/openlibertyapplication_types.go:
```diff
@@ -474,6 +560,7 @@ func (cr *OpenLibertyApplication) GetLabels() map[string]string {
 		"app.kubernetes.io/instance":   cr.Name,
 		"app.kubernetes.io/name":       cr.Name,
 		"app.kubernetes.io/managed-by": "open-liberty-operator",
+		"app.kubernetes.io/component":  "backend",
```
This is missing `part-of` and service binding label from before

edavidj committed:
Updates to CRD struct methods and gen comments

leochr commented:
@edavidj Thanks for the PR. 

- In Role and ClusterRole, is `pod/exec` needed for apiGroup `openliberty.io`? 
- test/e2e/openliberty_trace.go - Size changed from 1Gi to 1Mi - any particular reason? is 1Mi enough?

edavidj committed:
Generate changes and update e2e trace test storage

edavidj commented:
@leochr good catch, I had changed it for a local test with my own volume claim that was a smaller size. I've reverted that and gotten rid of the `pod/exec` under `openliberty.io` in the cluster role.

edavidj committed:
Align controller with runtime operator

edavidj committed:
Align types.go with runtime operator

edavidj committed:
Update utils_test to account for service reqs

edavidj committed:
Add resource constraint nil check to types

edavidj committed:
Update svc annotation assignment
